core/client/src/main/java/alluxio/client/block/stream/NettyPacketReader.java

```diff
@@ -0,0 +1,410 @@
+  private Throwable mPacketReaderException = null;
+  /** The condition is met when mPackets.size() > 0 or mPacketReaderException != null. */
+  private final Condition mNotEmptyOrFailed = mLock.newCondition();
+
+  /** The next pos to read. */
+  private long mPosToRead;
+  /** This is true only when an empty packet is received. */
+  private boolean mDone = false;
+
+  private boolean mClosed = false;
+
+  /**
+   * Creates an instance of {@link NettyPacketReader}. If this is used to read a block remotely, it
+   * requires the block to be locked beforehand and the lock ID is passed to this class.
+   *
+   * @param context the file system context
+   * @param address the netty data server network address
+   * @param id the block ID or UFS file ID
+   * @param offset the offset
+   * @param len the length to read
+   * @param lockId the lock ID
+   * @param sessionId the session ID
+   * @param type the request type (block or UFS file)
+   * @throws IOException if it fails to create the object
+   */
+  public NettyPacketReader(FileSystemContext context, InetSocketAddress address, long id,
+      long offset, long len, long lockId, long sessionId, Protocol.RequestType type)
+      throws IOException {
+    Preconditions.checkArgument(offset >= 0 && len > 0);
+
+    mContext = context;
+    mAddress = address;
+    mId = id;
+    mStart = offset;
+    mPosToRead = offset;
+    mBytesToRead = len;
+    mRequestType = type;
+
+    mChannel = context.acquireNettyChannel(address);
+
+    ChannelPipeline pipeline = mChannel.pipeline();
+    if (!(pipeline.last() instanceof RPCMessageDecoder)) {
+      throw new RuntimeException(String.format("Channel pipeline has unexpected handlers %s.",
+          pipeline.last().getClass().getCanonicalName()));
+    }
+    mChannel.pipeline().addLast(new PacketReadHandler());
+
+    Protocol.ReadRequest readRequest =
+        Protocol.ReadRequest.newBuilder().setId(id).setOffset(offset).setLength(len)
+            .setLockId(lockId).setSessionId(sessionId).setType(type).build();
+    mChannel.writeAndFlush(new RPCProtoMessage(readRequest))
+        .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
+  }
+
+  @Override
+  public long pos() {
+    return mPosToRead;
+  }
+
+  @Override
+  public ByteBuf readPacket() throws IOException {
+    Preconditions.checkState(!mClosed, "PacketReader is closed while reading packets.");
+    ByteBuf buf = null;
+    mLock.lock();
+    try {
+      while (true) {
+        if (mDone) {
+          return null;
+        }
+        if (mPacketReaderException != null) {
+          throw new IOException(mPacketReaderException);
+        }
+        buf = mPackets.poll();
+
+        // TODO(peis): Have a better criteria to resume so that we can have fewer state changes.
+        if (!tooManyPacketsPending()) {
+          resume();
+        }
+        if (buf == null) {
+          try {
+            if (!mNotEmptyOrFailed.await(READ_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
```

Comment:
do we need to unlock first before `await`? Otherwise, all `mNotEmptyOrFailed.signal();` is inside `mLock`, in other words, it seems we are awaiting here, no one will be able to signal.